curp/src/client_new/mod.rs

```diff
@@ -0,0 +1,168 @@
+/// Unary rpc client
+mod unary;
+
+use std::{collections::HashMap, sync::Arc};
+
+use async_trait::async_trait;
+use curp_external_api::cmd::Command;
+use futures::{stream::FuturesUnordered, StreamExt};
+use utils::config::ClientConfig;
+
+use crate::{
+    members::ServerId,
+    rpc::{
+        connect::ConnectApi, protocol_client::ProtocolClient, ConfChange, FetchClusterRequest,
+        FetchClusterResponse, Member, ReadState,
+    },
+};
+
+/// The response of propose command, deserialized from [`crate::rpc::ProposeResponse`] or
+/// [`crate::rpc::WaitSyncedResponse`].
+#[allow(type_alias_bounds)] // that's not bad
+type ProposeResponse<C: Command> = Result<(C::ER, Option<C::ASR>), C::Error>;
+
+/// `ClientApi`, a higher wrapper for `ConnectApi`, providing some methods for communicating to
+/// the whole curp cluster. Automatically discovery curp server to update it's quorum.
+#[async_trait]
+pub trait ClientApi<C: Command> {
+    /// The client error
+    type Error;
+
+    /// Get the local connection when the client is on the server node.
+    fn local_connect(&self) -> Option<Arc<dyn ConnectApi>>;
+
+    /// Send propose to the whole cluster, `use_fast_path` set to `false` to fallback into ordered
+    /// requests (event the requests are commutative).
+    async fn propose(
+        &self,
+        cmd: &C,
+        use_fast_path: bool,
+    ) -> Result<ProposeResponse<C>, Self::Error>;
+
+    /// Send propose configuration changes to the cluster
+    async fn propose_conf_change(
+        &self,
+        changes: Vec<ConfChange>,
+    ) -> Result<Vec<Member>, Self::Error>;
+
+    /// Send propose to shutdown cluster
+    async fn propose_shutdown(&self) -> Result<(), Self::Error>;
+
+    /// Send fetch read state from leader
+    async fn fetch_read_state(&self, cmd: &C) -> Result<ReadState, Self::Error>;
+
+    /// Send fetch cluster requests to all servers (That's because initially, we didn't
+    /// know who the leader is.)
+    /// Note: The fetched cluster may still be outdated if `linearizable` is false
+    async fn fetch_cluster(&self, linearizable: bool) -> Result<FetchClusterResponse, Self::Error>;
```

Comment:
Etcd just support that, which them called `serializable read`. And there is a field `linearizable` in `FetchClusterResponse` to enable quick cluster status retrieval (may return the stale data).

Reply:
In etcd, there is no FetchClusterRequest; only MemberListRequest exists. The MemberListResponse contains only Members, while FetchClusterResponse additionally includes leader_id and term. For the caller, once it obtains outdated leader_id and term, it will finally do another rpc call, which may lead to lower performance.

Reply:
> And there is a field linearizable in ~FetchClusterResponse~ to enable quick cluster status retrieval (may return the stale data).

It is `MemberListRequest`, I missed them up. FetchCluster is the only method for the curp client to correct the local cluster state and leader state. Retrying may result in an increase in RT for a particular RPC, but it will not affect the overall RT and QPS.

Reply:
if it doesn’t matter that the retrieved leader_id and term are outdated, you can simply read them directly from the local State without sending a FetchCluster Request. Reason has it that obtaining leader_id and term should always be linearizable, but MemberList is another story. 

IMO, we’d better move out the leader_id and the term from FetchClusterResponse if FetchClusterRequest must maintain linearizable field.  Alternatively, we can remove the linearizable from FetchClusterRequest to enforce that FetchCluster must be linearizable.

Reply:
No, Xline Cluster Server could send a non-linearizable `MemberListRequest`, we should implement it in FetchCluster

Reply:
@rogercloud

Reply:
After the reconstruction of ClusterServer, it is changed to full linearizable.